Add 5.12.0 release blog post (#164)
luanti-org/blog: _posts/2025-05-23-5.12.0-released.md

image: /static/blog/5.12.0/cover.webp
image_credit: 
forum_topic: https://forum.luanti.org/viewtopic.php?t=31686
tags:
  - release
---

<h2 class="sr-only">Introduction</h2>

Luanti 5.12.0 is here! Perfectly on schedule, we swear! Anyway, this time Luanti is sporting new input handling and controls, new item and entity visuals, better map performance, and more.

You can download Luanti from [our website](https://www.luanti.org/downloads/), check out the code on [GitHub](https://github.com/luanti-org/luanti/), or read the [full changelog](https://docs.luanti.org/changelog/#5110--5120).

### Table of Contents

- [SDL2 Window and Input Handling](#sdl2-window-and-input-handling)
- [Control Your Controls](#control-your-controls)
- [More Animation, Less Blur, and New Reviews](#more-animation-less-blur-and-new-reviews)
- [Camera Mode, Entity Visuals, and More](#camera-mode-entity-visuals-and-more)
- [World and Object Performance](#world-and-object-performance)
- [Other Highlights](#other-highlights)
- [What's Next?](#whats-next)

## SDL2 Window and Input Handling

> This section covers a plethora of outstanding issues, so don't skip this one— SDL is a bigger deal than you might think! Until this release, Luanti has relied on Irrlicht's custom window and input interfaces, which (much like a lot of old Irrlicht code) was _very_ outdated and buggy. [SDL](https://en.wikipedia.org/wiki/Simple_DirectMedia_Layer), on the other hand, has been consistently updated since 1998 and has been battle-tested by thousands of major applications over the years.

Reviewer: Wouldn't it be better to link to https://libsdl.org instead?

> This section covers a plethora of outstanding issues, so don't skip this one— SDL is a bigger deal than you might think! Until this release, Luanti has relied on Irrlicht's custom window and input interfaces, which (much like a lot of old Irrlicht code) was _very_ outdated and buggy. [SDL](https://en.wikipedia.org/wiki/Simple_DirectMedia_Layer), on the other hand, has been consistently updated since 1998 and has been battle-tested by thousands of major applications over the years.

Reviewer 2: ```suggestion
This section covers a plethora of outstanding issues, so don't skip this one— SDL is a bigger deal than you might think! Until this release, Luanti relied on Irrlicht's custom window and input interfaces, which (much like a lot of old Irrlicht code) was _very_ outdated and buggy. [SDL](https://en.wikipedia.org/wiki/Simple_DirectMedia_Layer), on the other hand, has been consistently updated since 1998 and has been battle-tested by thousands of major applications over the years.
```

Efforts to migrate to SDL2 have been ongoing for a few years now, and 5.12.0 is the first version to finally see those efforts come to fruition. The cross-platform wrappers SDL2 brings have enabled a bunch of cool stuff:
* Touchscreen controls are now supported on any touchscreen device (not just mobile) ([irr#262](https://github.com/minetest/irrlicht/pull/262), [#14146](https://github.com/luanti-org/luanti/pull/14146), [#14075](https://github.com/luanti-org/luanti/pull/14075), others)
* Keybinds now use [scancodes](https://en.wikipedia.org/wiki/Scancode), allowing non-QWERTY keyboards to work out of the box and fixing some platform-specific issues on macOS and others ([#14964](https://github.com/luanti-org/luanti/pull/14964))
* Support for high-DPI screens ([#14703](https://github.com/luanti-org/luanti/pull/14703))

> * Desktop [IME](https://en.wikipedia.org/wiki/Input_method) support for text input (helpful for CJK and other lanuages) ([#14942](https://github.com/luanti-org/luanti/pull/14942))

Reviewer: Nitpick since IME support for the X11 device was added in 2021 (minetest/irrlicht#23)

```suggestion
* Desktop [IME](https://en.wikipedia.org/wiki/Input_method) support for text input on all platforms (helpful for CJK and other lanuages) ([#14942](https://github.com/luanti-org/luanti/pull/14942))
```

The `node` visual type [has been introduced](https://github.com/luanti-org/luanti/pull/15683) to entities to represent any node (instead of hacking the `item` visual til it looks right).

```lua
initial_properties = {
    visual = "node",
    node = {
        name = "mymod:mynode",
        param2 = 42, -- It supports paramtypes too!
    }
}
```

Developers can now prevent the `Escape` key from closing their formspecs using [the `allow_close[true]` element](https://github.com/luanti-org/luanti/pull/15971). The form can still be closed with an exit button or with `Shift + Esc` in emergencies (so don't rely on your forms being locked open). The `try_quit` event field has also been introduced to watch for close attempts on locked forms.

As for notable fixes, custom hands now [properly override all tool capabilities](https://github.com/luanti-org/luanti/pull/15743). glTF models saw various fixes, most importantly local transforms [no longer affect meshes](https://github.com/luanti-org/luanti/pull/16086).

## World and Object Performance

> Alongside better world [loading](https://github.com/luanti-org/luanti/pull/16000) and [generation](https://github.com/luanti-org/luanti/pull/15856) performance, world storage [has been changed](https://github.com/luanti-org/luanti/pull/15768) to save block positions as separate integers instead of a hash. This simpler implementation is much easier to work with internally and for external tools. **Please note:** Worlds created with or migrated to 5.12.0 are **not backward-compatible.**

Reviewer 2: ```suggestion
Alongside better world [loading](https://github.com/luanti-org/luanti/pull/16000) and [generation](https://github.com/luanti-org/luanti/pull/15856) performance, world storage [has been changed](https://github.com/luanti-org/luanti/pull/15768) to save block positions as separate integers instead of a hash. This simpler implementation is much easier to work with internally and for external tools. **Please note:** Worlds created with or migrated to 5.12.0 are cannot be opened in older versions.
```

Reviewer 2: Nevermind this is not correct https://forum.luanti.org/viewtopic.php?p=444875#p444875
